trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/CegarLoopFactory.java

```diff
@@ -0,0 +1,350 @@
+					createAutomataAbstraction(services, root, errorLocs, predicateFactory, stateFactoryForRefinement,
+							witnessAutomaton),
+					root, csToolkit, predicateFactory, mPrefs, errorLocs, mPrefs.interpolation(),
+					mComputeHoareAnnotation, hoareAnnotationLocs, services, languageOperation, mTransitionClazz,
+					stateFactoryForRefinement);
+		}
+
+		if (mPrefs.interpolantAutomaton() == InterpolantAutomaton.TOTALINTERPOLATION) {
+			return new CegarLoopSWBnonRecursive<>(name,
+					createAutomataAbstraction(services, root, errorLocs, predicateFactory, stateFactoryForRefinement,
+							witnessAutomaton),
+					root, csToolkit, predicateFactory, mPrefs, errorLocs, mPrefs.interpolation(),
+					mComputeHoareAnnotation, hoareAnnotationLocs, services, mTransitionClazz,
+					stateFactoryForRefinement);
+		}
+
+		if (FORCE_FINITE_AUTOMATA_FOR_SEQUENTIAL_PROGRAMS && !IcfgUtils.isConcurrent(root)) {
+			return createFiniteAutomataCegarLoop(services, name, root, predicateFactory, errorLocs,
+					rawFloydHoareAutomataFromFile, hoareAnnotationLocs, stateFactoryForRefinement, witnessAutomaton);
+		}
+
+		switch (mPrefs.getAutomataTypeConcurrency()) {
+		case FINITE_AUTOMATA:
+			return createFiniteAutomataCegarLoop(services, name, root, predicateFactory, errorLocs,
+					rawFloydHoareAutomataFromFile, hoareAnnotationLocs, stateFactoryForRefinement, witnessAutomaton);
+		case PARTIAL_ORDER_FA:
+			requireNoReuse("POR-based analysis");
+			requireNoWitnesses(witnessAutomaton, "POR-based analysis");
+			return new PartialOrderCegarLoop<>(name,
+					createPartialOrderAbstraction(services, predicateFactory, stateFactoryForRefinement, root,
+							errorLocs),
+					root, csToolkit, predicateFactory, mPrefs, errorLocs, mPrefs.interpolation(), services,
+					mTransitionClazz, stateFactoryForRefinement);
+		case PETRI_NET:
+			requireNoReuse("Petri net-based analysis");
+			requireNoWitnesses(witnessAutomaton, "Petri net-based analysis");
+			return new CegarLoopForPetriNet<>(name,
+					createPetriAbstraction(services, predicateFactory, true, root, errorLocs), root, csToolkit,
+					predicateFactory, mPrefs, errorLocs, services, mTransitionClazz, stateFactoryForRefinement);
+		default:
+			throw new AssertionError("Unknown Setting: " + mPrefs.getAutomataTypeConcurrency());
+		}
+	}
+
+	private void requireNoReuse(final String analysis) {
+		if (mPrefs.getFloydHoareAutomataReuse() != FloydHoareAutomataReuse.NONE) {
+			throw new UnsupportedOperationException("Floyd/Hoare automaton reuse not supported for " + analysis);
+		}
+	}
+
+	private static void requireNoWitnesses(
+			final INwaOutgoingLetterAndTransitionProvider<WitnessEdge, WitnessNode> witnessAutomaton,
+			final String analysis) {
+		if (witnessAutomaton != null) {
+			throw new UnsupportedOperationException("Witness automata not supported for " + analysis);
+		}
+	}
+
+	private static PredicateFactory getPredicateFactory(final IUltimateServiceProvider services,
```

Comment:
Rename to constructPredicateFactory (get implies it is always the same instance)

Reply:
done